vllm/v1/core/block_pool.py

```diff
@@ -19,6 +19,101 @@
 logger = init_logger(__name__)
 
 
+class BlockHashToBlockMap:
+    """
+    Cache of blocks that are used for prefix caching.
+    """
+
+    def __init__(self):
+        # Caches blocks from hash directly to a block or multiple blocks
+        # (i.e. {block_hash: KVCacheBlocks})
+        # - Mostly block_hash maps to a single KVCacheBlock, and KVCacheBlocks
+        #   would simply be a KVCacheBlock.
+        # - Otherwise, KVCacheBlocks is a dict from {block_id: KVCacheBlock}
+        #
+        # A cached block is a full block with a block hash that can be used
+        # for prefix caching.
+        # The cached block may be used by running requests or in the
+        # free_block_queue that could potentially be evicted.
+        #
+        # NOTE #1: We currently don't de-duplicate the blocks in the cache,
+        # meaning that if a block becomes full and is cached, we don't check
+        # if there is already an identical block in the cache. This is because
+        # we want to make sure the allocated block IDs won't change so that
+        # block tables are append-only.
+        # NOTE #2: The union type is introduced in order to reduce GC costs
+        # from the inner dict.
+        self._cache: defaultdict[BlockHashWithGroupId,
+                                 Union[KVCacheBlock,
+                                       dict[int,
+                                            KVCacheBlock]]] = defaultdict(dict)
```

Comment:
This is a slightly cleaner supported syntax. However, I don't think this needs to be a `defaultdict`? Can it just be a regular `dict`?
```suggestion
        self._cache = defaultdict[BlockHashWithGroupId,
                                 Union[KVCacheBlock,
                                       dict[int, KVCacheBlock]]](dict)
```

Reply:
TIL. But change the dict seems fine with some unit test updates.